proposals/21-11-transformer-annotations.md

```diff
@@ -0,0 +1,259 @@
+option is set. If the source of the file points to a generator, we know that it is a generated resource.
+
+For local generators:
+
+```
+config.kubernetes.io/origin: |
+  path: path/to/generator.yaml
+```
+
+For remote generators:
+
+```
+config.kubernetes.io/origin: |
+  path: examples/multibases/base/generator.yaml
+  repo: https://github.com/kubernetes-sigs/kustomize
+  ref: v1.0.6 
+```
+
+
+### Retaining Transformation Data
+
+To retain the information about what transformers have acted on each resource, we can propose a new option 
+`transformerAnnotations` in the `buildMetadata` field of the kustomization:
+
+```yaml
+buildMetadata: [originAnnotations, transformerAnnotations]
+```
+
+When the `transformerAnnotations` option is set, kustomize will add annotations with information about what transformers 
+have acted on each resource. Transformers can be invoked either through various fields in the kustomization file 
+(e.g. the `replacements` field will invoke the ReplacementTransformer), or through the `transformers` field.
+
+#### Local transformers
+For transformers defined locally, kustomize will add the following annotation:
+
+```yaml
+config.kubernetes.io/transformations: |
+  - kind: TransformerKind 
+    path: filepathToTransformerConfig 
+    type: builtin/custom
+```
+
+`kind` is the kind from the transformer's GVK.
+
+`path` is the filepath to the file containing the transformer config, rooted at the directory upon which kustomize build was invoked.
+
+`type` has two legal values, custom or builtin, depending on whether the transformer is built into kustomize or is defined by the user.
+
+#### Remote transfoermers
+For transformers defined in a remote base, kustomize will an annotation like the following:
+
+```yaml
+config.kubernetes.io/transformations: |
+  - kind: NamespaceTransformer
+    path: examples/multibases/base/kustomization.yaml
+    repo: https://github.com/kubernetes-sigs/kustomize
+    ref: v1.0.6 
+    type: builtin
+```
+
+The only difference between the remote transformer and the local one is that the remote transformer will have `repo` and `ref` fields to store where the 
+transformer originated from. 
+ 
+A resource that has been through multiple transformers through several layers may be outputted by kustomize build with an annotation like the following:
+
+```yaml
+config.kubernetes.io/transformations: |
+  - kind: PrefixSuffixTransformer
+    path: ../overlays/kustomization.yaml
+    type: builtin
+  - kind: InflateHelmChart
+    path: inflatehelmchart.yaml
+    type: custom
+  - kind: NamespaceTransformer
+    path: examples/multibases/base/kustomization.yaml
+    repo: https://github.com/kubernetes-sigs/kustomize
+    ref: v1.0.6 
+    type: builtin
+```
+
+#### Wildcard match
+
+As a shortcut for all buildMetadata options, we can allow the user to use a wildcard match:
+
+```
+buildMetadata: [*]
```

Comment:
I am a bit hesitant with this. The potential problem I could see here is that if we ever add a new option to `buildMetadata`, a user might suddenly start seeing a bunch of new annotations that they weren't expecting.

The benefits though, are: 
1. The user can very easily enable all the debugging options available to them
2. Arguably if they are using a wildcard match, their output changing whenever there is a new option available is actually desirable. 

The alternative is to have `kustomize edit add buildMetadata *` add all the `buildMetadata` options to the kustomization explicitly, and not allow this sort of wildcard match in the kustomization itself. I'm open to either option.

Reply:
+1 to `kustomize edit`. This resolved my previous confusions on how users discover the kustomize syntax.

Reply:
sounds good! I have updated the proposal to only support wildcard match in `kustomize edit`.